### PR TITLE
fix(pyspark): set catalog and database with `USE` instead of pyspark api

### LINKED ISSUE #9604: bug:  unable to set database for table registered in databricks unity catalog using shared cluster
### What happened?

I am unable to read from a table in a databricks unity catalog when using a compute cluster with a shared access mode
When I try to load the table using

```python
spark = ...

import ibis
con = ibis.pyspark.connect(spark)

table = con.table("my_table", database=("catalog", "schema"))
```
I receive the following error

```
py4j.security.Py4JSecurityException: Method public void org.apache.spark.sql.internal.CatalogImpl.setCurrentCatalog(java.lang.String) is not whitelisted on class class org.apache.spark.sql.internal.CatalogImpl
```

From a quick google search, I think this is related to the fact that certain spark functions are not enabled on clusters with a shared access mode (e.g. [here](https://community.databricks.com/t5/data-governance/unity-catalog-spark-functions-throwing-py4jsecurityexception-org/td-p/7364)).

If I manually set the catalog and schema using SQL statements
```python
USE CATALOG catalog;
USE SCHEMA schema;
```
I am then able to load the table using

```python
table = con.table("my_table")
```

For further context, I have not encountered these issues when using a personal compute cluster that has a single-user access mode

### What version of ibis are you using?

9.1.0

### What backend(s) are you using, if any?

pyspark - 3.4.1

### Relevant log output

```sh
File /local_disk0/.ephemeral_nfs/envs/pythonEnv-6c76b61f-037f-47b2-b51d-bcb3b5d19e64/lib/python3.10/site-packages/ibis/backends/pyspark/__init__.py:237, in Backend._active_catalog_database(self, catalog, db)
    236 if catalog is not None:
--> 237     self._session.catalog.setCurrentCatalog(catalog)
    238 self._session.catalog.setCurrentDatabase(db)

File /databricks/spark/python/pyspark/instrumentation_utils.py:48, in _wrap_function.<locals>.wrapper(*args, **kwargs)
     47 try:
---> 48     res = func(*args, **kwargs)
     49     logger.log_success(
     50         module_name, class_name, function_name, time.perf_counter() - start, signature
     51     )

File /databricks/spark/python/pyspark/sql/catalog.py:122, in Catalog.setCurrentCatalog(self, catalogName)
    109 """Sets the current default catalog in this session.
    110 
    111 .. versionadded:: 3.4.0
   (...)
    120 >>> spark.catalog.setCurrentCatalog("spark_catalog")
    121 """
--> 122 return self._jcatalog.setCurrentCatalog(catalogName)

File /databricks/spark/python/lib/py4j-0.10.9.7-src.zip/py4j/java_gateway.py:1355, in JavaMember.__call__(self, *args)
   1354 answer = self.gateway_client.send_command(command)
-> 1355 return_value = get_return_value(
   1356     answer, self.gateway_client, self.target_id, self.name)
   1358 for temp_arg in temp_args:

File /databricks/spark/python/pyspark/errors/exceptions/captured.py:188, in capture_sql_exception.<locals>.deco(*a, **kw)
    187 try:
--> 188     return f(*a, **kw)
    189 except Py4JJavaError as e:

File /databricks/spark/python/lib/py4j-0.10.9.7-src.zip/py4j/protocol.py:330, in get_return_value(answer, gateway_client, target_id, name)
    329     else:
--> 330         raise Py4JError(
    331             "An error occurred while calling {0}{1}{2}. Trace:\n{3}\n".
    332             format(target_id, ".", name, value))
    333 else:
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by gforsyth**
## Description of changes

As reported in #9604, there is apparently a mismatch in permissioning structure
on Databricks, so that you can set a catalog and database (schema, ugh) with sql
by running `USE CATALOG`, but not via the pyspark api calls that (ostensibly) do
the same thing.

This, however, is not part of standard Spark SQL and is special
Databricks Spark SQL.

SO, we try the special databricks SQL, if that throws a parser error, we
fall back to using the pyspark API methods.

## Issues closed

Resolves #9604

### CODE REVIEW

**Review by cpcloud — APPROVED** [review]
Hard approve.

Is there any chance of us testing this?

**Comment by jstammers** [comment]
Thanks for the PR @gforsyth . I've just tested this and have the following error:
```python
ParseException: [[PARSE_SYNTAX_ERROR](https://docs.microsoft.com/azure/databricks/error-messages/error-classes#parse_syntax_error)] Syntax error at or near 'USE': extra input 'USE'.(line 2, pos 0)

== SQL ==
USE CATALOG hive_metastore;
USE SCHEMA default;
```
maybe they need to be executed as separate statements?

```python
    135     catalog = table_loc.catalog or None
    136     database = table_loc.db or None
--> 138 table_schema = self.get_schema(name, catalog=catalog, database=database)
    139 return ops.DatabaseTable(
    140     name,
    141     schema=table_schema,
    142     source=self,
    143     namespace=ops.Namespace(catalog=catalog, database=database),
    144 ).to_expr()
```

**Comment by gforsyth** [comment]
> maybe they need to be executed as separate statements?

Ahh, interesting, I bet you're right.  Let me try that.  

Thanks for testing it out!!

**Comment by gforsyth** [comment]
Hey @jstammers -- can you give this PR another shot when you have a chance?  Many thanks!

**Comment by jstammers** [comment]
That seems to work great for me. Thanks for resolving this so quickly!

**Comment by cpcloud** [comment]
We should probably have at least one test for this if possible. Is creating another catalog a huge pain?

**Comment by gforsyth** [comment]
Ok, this can probably be cleaned up at some point, but I think it's all working now.

File: ibis/backends/pyspark/__init__.py [hunk 1]
```diff
@@ -29,6 +29,11 @@
 from ibis.legacy.udf.vectorized import _coerce_to_series
 from ibis.util import deprecated
 
+try:
+    from pyspark.errors import ParseException as PySparkParseException
+except ImportError:
+    from pyspark.sql.utils import ParseException as PySparkParseException
```
Comment (cpcloud) [inline ibis/backends/pyspark/__init__.py:35]: image

**Review by cpcloud — APPROVED** [review]

**Pull request merged**